# Classes and objects

In [1]:
class PartyAnimal:
    x = 0

    def party(self):
        self.x = self.x + 1
        print("So far", self.x)

In [2]:
an = PartyAnimal()

In [3]:
an.party()
an.party()
an.party()

So far 1
So far 2
So far 3


In [4]:
print(type(an))

<class '__main__.PartyAnimal'>


In [5]:
print(dir(an))

['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'party', 'x']


**dir(object) returns the methods and attributes of the object**

In [6]:
an = 5

In [7]:
print(type(an))

<class 'int'>


In [8]:
an.party()

AttributeError: 'int' object has no attribute 'party'

In [9]:
an = PartyAnimal()

In [10]:
an.party()

So far 1


The PartyAnimal object __an__ is no longer referenced. <br/>
The Python garbage collector deallocates its memory

In [11]:
class PartyAnimal2:
    x = 0

    def __init__(self):
        print('I am constructed')

    def party(self) :
        self.x = self.x + 1
        print('So far',self.x)

    def __del__(self):
        print('I am destructed', self.x)

        
# an2 = PartyAnimal2()
# an2.party()
# an2.party()
# print(dir(an2))
# an2 = 42
# print('an contains', an)
# print(dir(an2))

In [12]:
an2 = PartyAnimal2()

I am constructed


In [13]:
an2.party()

So far 1


In [14]:
an2.party()

So far 2


In [15]:
print(dir(an2))

['__class__', '__del__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'party', 'x']


In [16]:
an2 = 42

I am destructed 2


The PartyAnimal object __an2__ is no longer referenced. Now you can see the automatic call to the onject destructor. 

In [ ]:
print('an contains', an, 'which is of type:' , type(an))

### dir() returns all the names in its current scope

In [17]:
def f1():
    x = 0
    y = 1
    z = 2
    print(dir())

In [18]:
f1()

['x', 'y', 'z']


## An Employee class (v1)

In [19]:
class Employee:
    ID = None
    first_name = None
    last_name = None
    salary = None
        
    def __init__(self, ID, fn, ln, sal):
        self.ID = ID
        self.first_name = fn
        self.last_name = ln
        self.salary = sal
        
    def promote(self, amount):
        self.salary += amount

In [20]:
e1 = Employee(101,'John', 'Doe', 10000)
e2 = Employee(102, 'Jane', 'Roe', 12000)
print(e1)
print(e2)

In [21]:
e2.promote(20000)

In [22]:
for e in [e1,e2]:
    print(e.ID, e.first_name, e.last_name, e.salary)

101 John Doe 10000
102 Jane Roe 32000


In [23]:
Employee(101,'John', 'Doe')

TypeError: Employee.__init__() missing 1 required positional argument: 'sal'

## An Employee class (v2)
Assume you wanted to keep track of the employees by putting them in a global list

In [25]:
employees = []

class Employee:
    ID = None
    first_name = None
    last_name = None
    salary = None
        
    def __init__(self, ID, fn, ln, sal):
        self.ID = ID
        self.first_name = fn
        self.last_name = ln
        self.salary = sal
        employees.append(self)

In [26]:
e3 = Employee(201,'John', 'Doe', 10000)
e4 = Employee(202, 'Jane', 'Roe', 12000)

print(employees, '\n')

for e in employees:
    print(e.ID, e.first_name, e.last_name, e.salary)

[<__main__.Employee object at 0x7f81daa7ba90>, <__main__.Employee object at 0x7f81daa7b8b0>] 

201 John Doe 10000
202 Jane Roe 12000


## An Employee class (v3)
Now you want the employe number to be assigned automatically (so that there are no errors)

In [27]:
employees = []

emp_id = 1000

class Employee:
    ID = None
    first_name = None
    last_name = None
    salary = None
        
    def __init__ (self, fn, ln, sal):  # no ID as argument
        global emp_id
        emp_id += 1
        self.ID = emp_id               # ID comes from the global emp_id
        self.first_name = fn
        self.last_name = ln
        self.salary = sal
        employees.append(self)
        

In [28]:
e5 = Employee('John', 'Doe', 10000)
e6 = Employee('Jane', 'Roe', 12000)
e7 = Employee('Jill', 'Moe', 15000)

for e in employees:
    print(e.ID, e.first_name, e.last_name, e.salary)

1001 John Doe 10000
1002 Jane Roe 12000
1003 Jill Moe 15000


In [30]:
print(employees)
print(emp_id)

[<__main__.Employee object at 0x7f81daa63700>, <__main__.Employee object at 0x7f81daa63c70>, <__main__.Employee object at 0x7f81daa63160>]
1003


In [32]:
for e in [e1,e2,e3,e4,e5,e6,e7]:
    print(e, e.ID, e.first_name, e.last_name, e.salary)

<__main__.Employee object at 0x7f81daaaa740> 101 John Doe 10000
<__main__.Employee object at 0x7f81daaaafe0> 102 Jane Roe 32000
<__main__.Employee object at 0x7f81daa7ba90> 201 John Doe 10000
<__main__.Employee object at 0x7f81daa7b8b0> 202 Jane Roe 12000
<__main__.Employee object at 0x7f81daa63700> 1001 John Doe 10000
<__main__.Employee object at 0x7f81daa63c70> 1002 Jane Roe 12000
<__main__.Employee object at 0x7f81daa63160> 1003 Jill Moe 15000


Note that objects with ids 101 and 1001 are distinc objects, even if the have the same name and salary. Same for 102 and 1002.

In [33]:
for e in [e1,e2,e3,e4,e5,e6,e7]:
    print(e)
    print(id(e), hex(id(e)))  # gives the memory location of e1 
    print(e.ID, e.first_name, e.last_name, e.salary)

140195696125760 0x7f81daaaa740
101 John Doe 10000
140195696127968 0x7f81daaaafe0
102 Jane Roe 32000
140195695934096 0x7f81daa7ba90
201 John Doe 10000
140195695933616 0x7f81daa7b8b0
202 Jane Roe 12000
140195695834880 0x7f81daa63700
1001 John Doe 10000
140195695836272 0x7f81daa63c70
1002 Jane Roe 12000
140195695833440 0x7f81daa63160
1003 Jill Moe 15000


(back to slides)

# Inheritance

In [34]:
class PartyAnimal:
    x = 0
    name = ""
    def __init__(self, nm):
        self.name = nm
        print(self.name, "constructed")
    def party(self):
        self.x = self.x + 1
        print(self.name, "party count", self.x)

class FootballFan(PartyAnimal):
    points = 0
    def touchdown(self):
        self.points = self.points + 7
        self.party()
        print(self.name, "points", self.points)

In [35]:
s = PartyAnimal("Sally")
s.party()

Sally constructed
Sally party count 1


In [36]:
j = FootballFan("Jim")
j.party()
print('points:', j.points)
j.touchdown()

Jim constructed
Jim party count 1
points: 0
Jim party count 2
Jim points 7


In [37]:
j.touchdown()

Jim party count 3
Jim points 14


### A Student is a subclass of Person
<img src="person-student.png" width="300" >

In [38]:
class Person:
    def __init__(self, name: str, age: int):
        self.name = name
        self.age = age
    def get_details(self) -> str:
        return f"Name: {self.name}, Age: {self.age}"

In [39]:
p1 = Person("John Doe", 30)
print(p1.get_details())

Name: John Doe, Age: 30


In [40]:
%%HTML
<iframe width="1100" height="500" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=class%20Person%3A%0A%20%20%20%20def%20__init__%28self,%20name%3A%20str,%20age%3A%20int%29%3A%0A%20%20%20%20%20%20%20%20self.name%20%3D%20name%0A%20%20%20%20%20%20%20%20self.age%20%3D%20age%0A%20%20%20%20def%20get_details%28self%29%20-%3E%20str%3A%0A%20%20%20%20%20%20%20%20return%20f%22Name%3A%20%7Bself.name%7D,%20Age%3A%20%7Bself.age%7D%22%0A%0Ap1%20%3D%20Person%28%22John%20Doe%22,%2030%29%0Aprint%28p1.get_details%28%29%29&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=311&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>


In [42]:
class Student(Person):
    def __init__(self, name: str, age: int, student_id: int, course: str):
        super().__init__(name, age)
        self.student_id = student_id
        self.course = course
        self.chapters_studied = 0
    def get_student_details(self) -> str:
	    person_details = super().get_details()
	    return f"{person_details}, Student ID: {self.student_id}, Course: {self.course}"
    def study(self, chapters: int = 1):
        """Increment the chapters_studied counter"""
        self.chapters_studied += chapters

In [45]:
s1 = Student("Alice Smith", 20, 1412325523, "DSCI510")
print(s1.get_student_details())
s1.study(7)
print(f'{s1.name} has studied {s1.chapters_studied} chapters')

Name: Alice Smith, Age: 20, Student ID: 1412325523, Course: DSCI510
Alice Smith has studied 7 chapters


In [44]:
%%HTML
<iframe width="1200" height="900" frameborder="0" src="https://pythontutor.com/iframe-embed.html#code=class%20Person%3A%0A%20%20%20%20def%20__init__%28self,%20name%3A%20str,%20age%3A%20int%29%3A%0A%20%20%20%20%20%20%20%20self.name%20%3D%20name%0A%20%20%20%20%20%20%20%20self.age%20%3D%20age%0A%20%20%20%20def%20get_details%28self%29%20-%3E%20str%3A%0A%20%20%20%20%20%20%20%20return%20f%22Name%3A%20%7Bself.name%7D,%20Age%3A%20%7Bself.age%7D%22%0A%20%20%20%20%20%20%20%20%0Aclass%20Student%28Person%29%3A%0A%20%20%20%20def%20__init__%28self,%20name%3A%20str,%20age%3A%20int,%20student_id%3A%20int,%20course%3A%20str%29%3A%0A%20%20%20%20%20%20%20%20super%28%29.__init__%28name,%20age%29%0A%20%20%20%20%20%20%20%20self.student_id%20%3D%20student_id%0A%20%20%20%20%20%20%20%20self.course%20%3D%20course%0A%20%20%20%20%20%20%20%20self.chapters_studied%20%3D%200%0A%20%20%20%20def%20get_student_details%28self%29%20-%3E%20str%3A%0A%20%20%20%20%20%20%20%20person_details%20%3D%20super%28%29.get_details%28%29%0A%20%20%20%20%20%20%20%20return%20f%22%7Bperson_details%7D,%20Student%20ID%3A%20%7Bself.student_id%7D,%20Course%3A%20%7Bself.course%7D%22%0A%20%20%20%20def%20study%28self,%20chapters%3A%20int%20%3D%201%29%3A%0A%20%20%20%20%20%20%20%20%22%22%22Increment%20the%20chapters_studied%20counter%22%22%22%0A%20%20%20%20%20%20%20%20self.chapters_studied%20%2B%3D%20chapters%0A%20%20%20%20%20%20%20%20%0As1%20%3D%20Student%28%22Alice%20Smith%22,%2020,%201412325523,%20%22DSCI510%22%29%0Aprint%28s1.get_student_details%28%29%29%0As1.study%287%29%0Aprint%28f'%7Bs1.name%7D%20has%20studied%20%7Bs1.chapters_studied%7D%20chapters'%29&codeDivHeight=400&codeDivWidth=350&cumulative=false&curInstr=0&heapPrimitives=nevernest&origin=opt-frontend.js&py=311&rawInputLstJSON=%5B%5D&textReferences=false"> </iframe>


In [46]:
s1.study()
print(f'{s1.name} has studied {s1.chapters_studied} chapters')

Alice Smith has studied 8 chapters
